src/Cli/dotnet/commands/dotnet-run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
+    {
+        if (directives.Length == 0)
+        {
+            return null;
+        }
+
+        Debug.Assert(directives.OrderBy(d => d.Span.Start).SequenceEqual(directives), "Directives should be ordered by source location.");
+
+        for (int i = directives.Length - 1; i >= 0; i--)
+        {
+            var directive = directives[i];
+            text = text.Replace(directive.Span, string.Empty);
+        }
+
+        return text;
+    }
+
+    public static void RemoveDirectivesFromFile(ImmutableArray<CSharpDirective> directives, SourceText text, string filePath)
+    {
+        if (RemoveDirectivesFromFile(directives, text) is { } modifiedText)
         {
-            projectRoot = ProjectRootElement.Create(xmlReader, projectCollection);
+            using var stream = File.Open(filePath, FileMode.Create, FileAccess.Write);
+            using var writer = new StreamWriter(stream, Encoding.UTF8);
+            modifiedText.Write(writer);
         }
-        projectRoot.AddItem(itemType: "Compile", include: EntryPointFileFullPath);
-        projectRoot.FullPath = projectFileFullPath;
-        return projectRoot;
     }
 
     public static bool IsValidEntryPointPath(string entryPointFilePath)
     {
         return entryPointFilePath.EndsWith(".cs", StringComparison.OrdinalIgnoreCase) && File.Exists(entryPointFilePath);
     }
 }
+
+internal readonly record struct SourceFile(string Path, SourceText Text)
+{
+    public string GetLocationString(TextSpan span)
+    {
+        var positionSpan = new FileLinePositionSpan(Path, Text.Lines.GetLinePositionSpan(span));
+        return $"{positionSpan.Path}:{positionSpan.StartLinePosition.Line + 1}";
+    }
+}
+
+internal static partial class Patterns
+{
+    [GeneratedRegex("""^\s*#:\s*(\w*)\s*(.*?)\s*$""")]
+    public static partial Regex Directive { get; }
+
+    [GeneratedRegex("""^\s*#!.*$""")]
+    public static partial Regex Shebang { get; }
+}
+
+/// <summary>
+/// Represents a C# directive starting with <c>#:</c>. Those are ignored by the language but recognized by us.
+/// </summary>
+internal abstract record CSharpDirective
+{
+    private static readonly SearchValues<char> s_separators = SearchValues.Create('/', '=');
+
+    private CSharpDirective() { }
+
+    /// <summary>
+    /// Span of the full line including the trailing line break.
+    /// </summary>
+    public required TextSpan Span { get; init; }
+
+    public static CSharpDirective Parse(SourceFile sourceFile, TextSpan span, string name, string value)
+    {
+        return name switch
+        {
+            "sdk" => Sdk.Parse(sourceFile, span, name, value),
+            "property" => Property.Parse(sourceFile, span, name, value),
+            "package" => Package.Parse(sourceFile, span, name, value),
+            _ => throw new GracefulException(LocalizableStrings.UnrecognizedDirective, name, sourceFile.GetLocationString(span)),
```

Comment:
There are other directives that can just stay in the C# file, so I think it'd be best to just not make a directive in this case (and not remove it from the C# file)

Reply:
This is only handling `#:` directives and there shouldn't be any other than those three. (I.e., this is not handling existing C# directives like `#pragma`.) We want to error on unknown `#:` directives to reserve them for future use.